File: cf_sdk/src/main/java/com/example/cf_sdk/logic/changebanksdk/util/ChangebankError.java
package com.example.cf_sdk.logic.changebanksdk.util;

import io.reactivex.Completable;
import io.reactivex.CompletableSource;
import io.reactivex.CompletableTransformer;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.SingleTransformer;

/**
 *
 * Base class for ErrorManager implementation.
 */

public interface ChangebankError<T> extends SingleTransformer<T, T>,
        CompletableTransformer,
        ObservableTransformer<T, T> {
    @Override
    SingleSource<T> apply(Single<T> upstream);

    @Override
    CompletableSource apply(Completable upstream);

    @Override
    ObservableSource<T> apply(Observable<T> upstream);
}

File: cf_sdk/src/main/java/com/example/cf_sdk/changebankapi/util/ChangebankError.kt
package com.example.cf_sdk.changebankapi.util

import io.reactivex.Completable
import io.reactivex.CompletableSource
import io.reactivex.CompletableTransformer
import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.ObservableTransformer
import io.reactivex.Single
import io.reactivex.SingleSource
import io.reactivex.SingleTransformer

/**
 *
 * Base class for ErrorManager implementation.
 */
interface ChangebankError<T> : SingleTransformer<T, T>, CompletableTransformer,
    ObservableTransformer<T, T> {
    override fun apply(upstream: Single<T>): SingleSource<T>
    override fun apply(upstream: Completable): CompletableSource
    override fun apply(upstream: Observable<T>): ObservableSource<T>
}